cli/diagnostics.rs

```diff
@@ -0,0 +1,641 @@
+  /// The highlight is a warning. This will place yellow carets under the
+  /// highlight.
+  Warning,
+  #[allow(dead_code)]
+  /// The highlight shows code additions. This will place green + signs under
+  /// the highlight and will highlight the code in green.
+  Addition,
+  /// The highlight shows a hint. This will place blue dashes under the
+  /// highlight.
+  Hint,
+}
+
+impl DiagnosticSnippetHighlightStyle {
+  fn style_underline(
+    &self,
+    s: impl std::fmt::Display,
+  ) -> impl std::fmt::Display {
+    match self {
+      DiagnosticSnippetHighlightStyle::Error => colors::red_bold(s),
+      DiagnosticSnippetHighlightStyle::Warning => colors::yellow_bold(s),
+      DiagnosticSnippetHighlightStyle::Addition => colors::green_bold(s),
+      DiagnosticSnippetHighlightStyle::Hint => colors::intense_blue(s),
+    }
+  }
+
+  fn underline_char(&self) -> char {
+    match self {
+      DiagnosticSnippetHighlightStyle::Error => '^',
+      DiagnosticSnippetHighlightStyle::Warning => '^',
+      DiagnosticSnippetHighlightStyle::Addition => '+',
+      DiagnosticSnippetHighlightStyle::Hint => '-',
+    }
+  }
+}
+
+pub enum DiagnosticSnippetSource<'a> {
+  /// The specifier of the module that should be displayed in this snippet. The
+  /// contents of the file will be retrieved from the `SourceTextStore`.
+  Specifier(Cow<'a, ModuleSpecifier>),
+  #[allow(dead_code)]
+  /// The source text that should be displayed in this snippet.
+  ///
+  /// This should be used if the text of the snippet is not available in the
+  /// `SourceTextStore`.
+  SourceTextInfo(Cow<'a, deno_ast::SourceTextInfo>),
+}
+
+impl<'a> DiagnosticSnippetSource<'a> {
+  fn to_source_text_info(
+    &self,
+    sources: &'a dyn SourceTextStore,
+  ) -> Cow<'a, SourceTextInfo> {
+    match self {
+      DiagnosticSnippetSource::Specifier(specifier) => {
+        sources.get_source_text(specifier).expect(
+          "source text should be in the cache if snippet source is a specifier",
+        )
+      }
+      DiagnosticSnippetSource::SourceTextInfo(info) => info.clone(),
+    }
+  }
+}
+
+/// Returns the text of the line with the given number.
+fn line_text(source: &SourceTextInfo, line_number: usize) -> &str {
+  source.line_text(line_number - 1)
+}
+
+/// Returns the text of the line that contains the given position, split at the
+/// given position.
+fn line_text_split(
+  source: &SourceTextInfo,
+  pos: DiagnosticSourcePos,
+) -> (&str, &str) {
+  let pos = pos.pos(source);
+  let line_index = source.line_index(pos);
+  let line_start_pos = source.line_start(line_index);
+  let line_end_pos = source.line_end(line_index);
+  let before = source.range_text(&SourceRange::new(line_start_pos, pos));
+  let after = source.range_text(&SourceRange::new(pos, line_end_pos));
+  (before, after)
+}
+
+/// Returns the text of the line that contains the given positions, split at the
+/// given positions.
+///
+/// If the positions are on different lines, this will panic.
```

Comment:
Is this okay? Maybe it should return an option and not panic?

Reply:
The caller ensures they are on the same line already.